Upload pdf document to the platform

As part of the platform, we need to be able to upload all versions of the document in the platform, for tracking. The documents uploaded will be saved to S3 bucket. The documents uploaded will not be parsed, just tracked.

@RaduCStefanescu could you review, please?
I'd like to be sure that what I wrote is actually correct.
I haven't written tests yet. The code compiles, but I get
```
[ERROR] Errors:
[ERROR]   AuthControllerIntegrationTest.login » IllegalState Failed to load ApplicationC...
[ERROR]   AuthControllerIntegrationTest.signUp » IllegalState Failed to load Application...
[ERROR]   CommentControllerIntegrationTest.create » IllegalState Failed to load Applicat...
[ERROR]   CommentControllerIntegrationTest.createReply » IllegalState Failed to load App...
[ERROR]   CommentControllerIntegrationTest.deleteComment » IllegalState Failed to load A...
[ERROR]   CommentControllerIntegrationTest.findAll » IllegalState Failed to load Applica...
[ERROR]   CommentControllerIntegrationTest.findAllReplies » IllegalState Failed to load ...
[ERROR]   CommentControllerIntegrationTest.update » IllegalState Failed to load Applicat...
[ERROR]   DocumentControllerIntegrationTest.deleteDocument » IllegalState Failed to load...
[ERROR]   DocumentControllerIntegrationTest.deleteDocumentNotFound » IllegalState Failed...
[ERROR]   DocumentControllerIntegrationTest.getDocument » IllegalState Failed to load Ap...
[ERROR]   DocumentControllerIntegrationTest.getDocumentCommentsNumber » IllegalState Fai...
[ERROR]   DocumentControllerIntegrationTest.listDocuments » IllegalState Failed to load ...
[ERROR]   DocumentControllerIntegrationTest.saveDocument » IllegalState Failed to load A...
[ERROR]   DocumentControllerIntegrationTest.saveDocumentDuplicatedFilePath » IllegalState
[ERROR]   DocumentControllerIntegrationTest.saveDocumentDuplicatedNumber » IllegalState ...
[ERROR]   DocumentControllerIntegrationTest.testGetDocumentConsolidatedNotFound » IllegalState
[ERROR]   DocumentControllerIntegrationTest.testGetDocumentNotFound » IllegalState Faile...
[ERROR]   DocumentControllerIntegrationTest.testGetDocumentsPagination » IllegalState Fa...
[ERROR]   DocumentControllerIntegrationTest.testGetDocumentsPaginationAndSorting » IllegalState
[ERROR]   DocumentControllerIntegrationTest.testGetDocumentsPaginationForNonExistingPage » IllegalState
[ERROR]   DocumentControllerIntegrationTest.testGetDocumentsSorting » IllegalState Faile...
[ERROR]   DocumentControllerIntegrationTest.testGetDocumentsSortingByNonExistentField » IllegalState
[ERROR]   FileControllerIntegrationTest.deleteFile » IllegalState Failed to load Applica...
[ERROR]   FileControllerIntegrationTest.saveFile » IllegalState Failed to load Applicati...
[ERROR]   UserControllerIntegrationTest.deleteUser » IllegalState Failed to load Applica...
[ERROR]   UserControllerIntegrationTest.extractUserFromCopy » IllegalState Failed to loa...
[ERROR]   UserControllerIntegrationTest.extractUsers » IllegalState Failed to load Appli...
[ERROR]   UserControllerIntegrationTest.getUser » IllegalState Failed to load Applicatio...
[ERROR]   UserControllerIntegrationTest.list » IllegalState Failed to load ApplicationCo...
[ERROR]   UserControllerIntegrationTest.saveUser » IllegalState Failed to load Applicati...
[ERROR]   UserControllerIntegrationTest.saveUserEmailException » IllegalState Failed to ...
[ERROR]   UserControllerIntegrationTest.saveUsers » IllegalState Failed to load Applicat...
```
I assume the errors are thrown because of the changes I made to `DocumentConsolidated`.

--- a/src/main/java/com/code4ro/legalconsultation/controller/DocumentController.java
+++ b/src/main/java/com/code4ro/legalconsultation/controller/DocumentController.java
@@ -4,7 +4,6 @@ import com.code4ro.legalconsultation.model.DocumentUserAssignmentDto;
 import com.code4ro.legalconsultation.model.dto.*;
 import com.code4ro.legalconsultation.model.persistence.DocumentConsolidated;
 import com.code4ro.legalconsultation.model.persistence.DocumentMetadata;
-import com.code4ro.legalconsultation.model.persistence.PdfHandle;
 import com.code4ro.legalconsultation.service.api.DocumentService;
 import io.swagger.annotations.ApiOperation;
 import io.swagger.annotations.ApiParam;
@@ -105,7 +104,7 @@ public class DocumentController {
             produces = MediaType.APPLICATION_JSON_VALUE,
             response = PdfHandleDto.class)
     @PostMapping("/{id}/pdf")
-    public ResponseEntity<PdfHandle> uploadPdf(@ApiParam(value = "Id of the document") @PathVariable UUID id,
+    public ResponseEntity<PdfHandleDto> uploadPdf(@ApiParam(value = "Id of the document") @PathVariable UUID id,
                                                @ApiParam(value = "State of the pdf document") @RequestParam String state,
                                                @ApiParam(value = "The pdf document") @RequestBody MultipartFile file) {
         return ResponseEntity.ok(documentService.addPdf(id, state, file));

--- a/src/main/java/com/code4ro/legalconsultation/controller/DocumentController.java
+++ b/src/main/java/com/code4ro/legalconsultation/controller/DocumentController.java
@@ -4,6 +4,7 @@ import com.code4ro.legalconsultation.model.DocumentUserAssignmentDto;
 import com.code4ro.legalconsultation.model.dto.*;
 import com.code4ro.legalconsultation.model.persistence.DocumentConsolidated;
 import com.code4ro.legalconsultation.model.persistence.DocumentMetadata;
+import com.code4ro.legalconsultation.model.persistence.PdfHandle;
 import com.code4ro.legalconsultation.service.api.DocumentService;
 import io.swagger.annotations.ApiOperation;
 import io.swagger.annotations.ApiParam;
@@ -14,6 +15,7 @@ import org.springframework.http.HttpStatus;
 import org.springframework.http.MediaType;
 import org.springframework.http.ResponseEntity;
 import org.springframework.web.bind.annotation.*;
+import org.springframework.web.multipart.MultipartFile;
 
 import javax.validation.Valid;
 import java.util.List;
@@ -96,5 +98,16 @@ public class DocumentController {
     public ResponseEntity<List<UserDto>> getAssignedUsers(@ApiParam(value = "Id of the document") @PathVariable("id") UUID id) {
         final List<UserDto> assignedUsers = documentService.getAssignedUsers(id);
         return ResponseEntity.ok(assignedUsers);
+    }
+    
+    @ApiOperation(value = "Upload a pdf to this document",
+            consumes = MediaType.APPLICATION_PDF_VALUE,
+            produces = MediaType.APPLICATION_JSON_VALUE,
+            response = PdfHandleDto.class)
+    @PostMapping("/{id}/pdf")
+    public ResponseEntity<PdfHandle> uploadPdf(@ApiParam(value = "Id of the document") @PathVariable UUID id,
+                                               @ApiParam(value = "State of the pdf document") @RequestParam String state,
+                                               @ApiParam(value = "The pdf document") @RequestBody MultipartFile file) {
+        return ResponseEntity.ok(documentService.addPdf(id, state, file));
     }
 }

--- a/src/main/java/com/code4ro/legalconsultation/converters/PdfHandleMapper.java
+++ b/src/main/java/com/code4ro/legalconsultation/converters/PdfHandleMapper.java
@@ -1,0 +1,11 @@
+package com.code4ro.legalconsultation.converters;
+
+import com.code4ro.legalconsultation.model.dto.PdfHandleDto;
+import com.code4ro.legalconsultation.model.persistence.PdfHandle;
+import org.mapstruct.Mapper;
+
+@Mapper(componentModel = "spring")
+public interface PdfHandleMapper {
+    PdfHandle map(PdfHandleDto pdfHandleDto);
+    PdfHandleDto map(PdfHandle pdfHandle);
+}

--- a/src/main/java/com/code4ro/legalconsultation/model/dto/PdfHandleDto.java
+++ b/src/main/java/com/code4ro/legalconsultation/model/dto/PdfHandleDto.java
@@ -1,0 +1,16 @@
+package com.code4ro.legalconsultation.model.dto;
+
+import lombok.Getter;
+import lombok.Setter;
+
+import java.net.URI;
+import java.time.Instant;
+
+@Getter
+@Setter
+public class PdfHandleDto extends BaseEntityDto {
+    private String state;
+    private URI uri;
+    private Instant timestamp;
+    private Integer hash;
+}

--- a/src/main/java/com/code4ro/legalconsultation/model/persistence/PdfHandle.java
+++ b/src/main/java/com/code4ro/legalconsultation/model/persistence/PdfHandle.java
@@ -1,0 +1,35 @@
+package com.code4ro.legalconsultation.model.persistence;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+import javax.persistence.*;
+import java.net.URI;
+import java.time.Instant;
+
+@Entity
+@Table(name = "pdf_handle")
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+public class PdfHandle extends BaseEntity {
+    @Column(nullable = false)
+    private String state;
+
+    @Column(nullable = false, unique = true)
+    private URI uri;
+
+    @Column(nullable = false)
+    private Instant timestamp;
+
+    @Column(nullable = false, unique = true)
+    private Integer hash;
+
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "consolidated_document_id", nullable = false)
+    private DocumentConsolidated owner;
+}

--- a/src/main/java/com/code4ro/legalconsultation/repository/PdfHandleRepository.java
+++ b/src/main/java/com/code4ro/legalconsultation/repository/PdfHandleRepository.java
@@ -1,0 +1,16 @@
+package com.code4ro.legalconsultation.repository;
+
+import com.code4ro.legalconsultation.model.persistence.PdfHandle;
+import lombok.NonNull;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+import java.util.UUID;
+
+@Repository
+public interface PdfHandleRepository extends JpaRepository<PdfHandle, UUID> {
+    PdfHandle findByHash(Integer hash);
+    PdfHandle findByHashAndState(Integer hash, String state);
+    Boolean existsByHash(@NonNull Integer hash);
+    Boolean existsByHashAndState(@NonNull Integer hash, @NonNull String state);
+}

--- a/src/main/java/com/code4ro/legalconsultation/service/api/DocumentService.java
+++ b/src/main/java/com/code4ro/legalconsultation/service/api/DocumentService.java
@@ -6,9 +6,11 @@ import com.code4ro.legalconsultation.model.dto.DocumentViewDto;
 import com.code4ro.legalconsultation.model.dto.UserDto;
 import com.code4ro.legalconsultation.model.persistence.DocumentConsolidated;
 import com.code4ro.legalconsultation.model.persistence.DocumentMetadata;
+import com.code4ro.legalconsultation.model.persistence.PdfHandle;
 import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Service;
+import org.springframework.web.multipart.MultipartFile;
 
 import java.util.List;
 import java.util.Set;
@@ -39,4 +41,6 @@ public interface DocumentService {
     void assignUsers(final UUID id, final Set<UUID> userIds);
 
     List<UserDto> getAssignedUsers(final UUID id);
+
+    PdfHandle addPdf(final UUID id, final String state, final MultipartFile file);
 }

--- a/src/main/java/com/code4ro/legalconsultation/service/api/DocumentService.java
+++ b/src/main/java/com/code4ro/legalconsultation/service/api/DocumentService.java
@@ -1,12 +1,8 @@
 package com.code4ro.legalconsultation.service.api;
 
-import com.code4ro.legalconsultation.model.dto.DocumentConsolidatedDto;
-import com.code4ro.legalconsultation.model.dto.DocumentMetadataDto;
-import com.code4ro.legalconsultation.model.dto.DocumentViewDto;
-import com.code4ro.legalconsultation.model.dto.UserDto;
+import com.code4ro.legalconsultation.model.dto.*;
 import com.code4ro.legalconsultation.model.persistence.DocumentConsolidated;
 import com.code4ro.legalconsultation.model.persistence.DocumentMetadata;
-import com.code4ro.legalconsultation.model.persistence.PdfHandle;
 import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Service;
@@ -42,5 +38,5 @@ public interface DocumentService {
 
     List<UserDto> getAssignedUsers(final UUID id);
 
-    PdfHandle addPdf(final UUID id, final String state, final MultipartFile file);
+    PdfHandleDto addPdf(final UUID id, final String state, final MultipartFile file);
 }

--- a/src/main/java/com/code4ro/legalconsultation/service/api/PDFService.java
+++ b/src/main/java/com/code4ro/legalconsultation/service/api/PDFService.java
@@ -1,5 +1,11 @@
 package com.code4ro.legalconsultation.service.api;
 
+import com.code4ro.legalconsultation.model.persistence.DocumentConsolidated;
+import com.code4ro.legalconsultation.model.persistence.PdfHandle;
+import org.springframework.web.multipart.MultipartFile;
+
 public interface PDFService {
     String readAsString(byte[] file);
+
+    PdfHandle createPdf(DocumentConsolidated documentConsolidatedId, String state, MultipartFile file);
 }

--- a/src/main/java/com/code4ro/legalconsultation/service/impl/DocumentServiceImpl.java
+++ b/src/main/java/com/code4ro/legalconsultation/service/impl/DocumentServiceImpl.java
@@ -15,6 +15,7 @@ import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Service;
 import org.springframework.transaction.annotation.Transactional;
+import org.springframework.web.multipart.MultipartFile;
 
 import java.math.BigInteger;
 import java.util.List;
@@ -134,5 +135,12 @@ public class DocumentServiceImpl implements DocumentService {
         final List<User> assignedUsers = documentConsolidated.getAssignedUsers();
 
         return assignedUsers.stream().map(userMapperService::map).collect(Collectors.toList());
+    }
+
+    @Override
+    public PdfHandle addPdf(final UUID id, final String state, final MultipartFile file) {
+        final DocumentConsolidated documentConsolidated = documentConsolidatedService.getEntity(id);
+
+        return pdfService.createPdf(documentConsolidated, state, file);
     }
 }

--- a/src/main/java/com/code4ro/legalconsultation/service/impl/DocumentServiceImpl.java
+++ b/src/main/java/com/code4ro/legalconsultation/service/impl/DocumentServiceImpl.java
@@ -1,11 +1,9 @@
 package com.code4ro.legalconsultation.service.impl;
 
 import com.code4ro.legalconsultation.converters.DocumentConsolidatedMapper;
+import com.code4ro.legalconsultation.converters.PdfHandleMapper;
 import com.code4ro.legalconsultation.converters.UserMapper;
-import com.code4ro.legalconsultation.model.dto.DocumentConsolidatedDto;
-import com.code4ro.legalconsultation.model.dto.DocumentMetadataDto;
-import com.code4ro.legalconsultation.model.dto.DocumentViewDto;
-import com.code4ro.legalconsultation.model.dto.UserDto;
+import com.code4ro.legalconsultation.model.dto.*;
 import com.code4ro.legalconsultation.model.persistence.*;
 import com.code4ro.legalconsultation.service.api.*;
 import org.slf4j.Logger;
@@ -35,6 +33,7 @@ public class DocumentServiceImpl implements DocumentService {
     private final UserMapper userMapperService;
     private final DocumentConsolidatedMapper documentConsolidatedMapper;
     private final CommentService commentService;
+    private final PdfHandleMapper pdfHandleMapper;
 
     @Autowired
     public DocumentServiceImpl(final DocumentConsolidatedService documentConsolidatedService,
@@ -45,7 +44,8 @@ public class DocumentServiceImpl implements DocumentService {
                                final UserService userService,
                                final UserMapper userMapperService,
                                final DocumentConsolidatedMapper documentConsolidatedMapper,
-                               final CommentService commentService) {
+                               final CommentService commentService,
+                               final PdfHandleMapper pdfHandleMapper) {
         this.documentConsolidatedService = documentConsolidatedService;
         this.documentMetadataService = documentMetadataService;
         this.pdfService = pdfService;
@@ -55,6 +55,7 @@ public class DocumentServiceImpl implements DocumentService {
         this.userMapperService = userMapperService;
         this.documentConsolidatedMapper = documentConsolidatedMapper;
         this.commentService = commentService;
+        this.pdfHandleMapper = pdfHandleMapper;
     }
 
     @Transactional(readOnly = true)
@@ -138,9 +139,9 @@ public class DocumentServiceImpl implements DocumentService {
     }
 
     @Override
-    public PdfHandle addPdf(final UUID id, final String state, final MultipartFile file) {
+    public PdfHandleDto addPdf(final UUID id, final String state, final MultipartFile file) {
         final DocumentConsolidated documentConsolidated = documentConsolidatedService.getEntity(id);
 
-        return pdfService.createPdf(documentConsolidated, state, file);
+        return pdfHandleMapper.map(pdfService.createPdf(documentConsolidated, state, file));
     }
 }

--- a/src/main/java/com/code4ro/legalconsultation/service/impl/pdf/PDFServiceImpl.java
+++ b/src/main/java/com/code4ro/legalconsultation/service/impl/pdf/PDFServiceImpl.java
@@ -1,26 +1,46 @@
 package com.code4ro.legalconsultation.service.impl.pdf;
 
 import com.code4ro.legalconsultation.common.exceptions.LegalValidationException;
+import com.code4ro.legalconsultation.converters.PdfHandleMapper;
+import com.code4ro.legalconsultation.model.persistence.DocumentConsolidated;
+import com.code4ro.legalconsultation.model.persistence.PdfHandle;
+import com.code4ro.legalconsultation.repository.PdfHandleRepository;
 import com.code4ro.legalconsultation.service.api.PDFService;
+import com.code4ro.legalconsultation.service.api.StorageApi;
 import com.code4ro.legalconsultation.service.impl.pdf.reader.BasicOARPdfReader;
 import com.code4ro.legalconsultation.service.impl.pdf.reader.PDFReader;
+import lombok.NonNull;
 import lombok.extern.slf4j.Slf4j;
 import org.apache.pdfbox.pdmodel.PDDocument;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.http.HttpStatus;
 import org.springframework.stereotype.Service;
+import org.springframework.web.multipart.MultipartFile;
 
 import java.io.IOException;
+import java.net.URI;
+import java.time.Instant;
+import java.util.Collections;
 
 @Service
 @Slf4j
 public class PDFServiceImpl implements PDFService {
 
     private final BasicOARPdfReader basicOARPdfReader;
+    
+    private final StorageApi storageApi;
+    private final PdfHandleMapper pdfHandleMapper;
+    private final PdfHandleRepository pdfHandleRepository;
 
     @Autowired
-    public PDFServiceImpl(final BasicOARPdfReader basicOARPdfReader) {
+    public PDFServiceImpl(final BasicOARPdfReader basicOARPdfReader,
+                          final StorageApi storageApi,
+                          final PdfHandleMapper pdfHandleMapper,
+                          final PdfHandleRepository pdfHandleRepository) {
         this.basicOARPdfReader = basicOARPdfReader;
+        this.storageApi = storageApi;
+        this.pdfHandleMapper = pdfHandleMapper;
+        this.pdfHandleRepository = pdfHandleRepository;
     }
 
     @Override
@@ -35,5 +55,37 @@ public class PDFServiceImpl implements PDFService {
             log.warn("Exception while parsing PDF file", e);
             throw new LegalValidationException("document.parse.pdf.failed", HttpStatus.BAD_REQUEST);
         }
+    }
+
+    @Override
+    public PdfHandle createPdf(@NonNull DocumentConsolidated owner, String state, MultipartFile file) {
+        Integer hash = file.hashCode();
+        if (pdfHandleRepository.existsByHash(hash)) {
+            if (pdfHandleRepository.existsByHashAndState(hash, state)) {
+                return pdfHandleRepository.findByHashAndState(hash, state);
+            }
+            throw new LegalValidationException("document.pdf.already_exists",
+                    Collections.singletonList(pdfHandleRepository.findByHash(hash).getId().toString()),
+                    HttpStatus.CONFLICT);
+        }
+
+        final String uriString;
+        try {
+            uriString = storageApi.storeFile(file);
+        } catch (Exception e) {
+            log.error("failed to save the pdf file", e);
+            throw new LegalValidationException("document.pdf.upload.failed",
+                    Collections.singletonList(e.toString()),
+                    HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+
+        PdfHandle pdfHandle = new PdfHandle();
+        pdfHandle.setHash(hash);
+        pdfHandle.setOwner(owner);
+        pdfHandle.setState(state);
+        pdfHandle.setUri(URI.create(uriString));
+        pdfHandle.setTimestamp(Instant.now());
+
+        return pdfHandleRepository.save(pdfHandle);
     }
 }

--- a/src/test/java/com/code4ro/legalconsultation/controller/DocumentControllerIntegrationTest.java
+++ b/src/test/java/com/code4ro/legalconsultation/controller/DocumentControllerIntegrationTest.java
@@ -10,6 +10,7 @@ import com.code4ro.legalconsultation.model.persistence.DocumentNode;
 import com.code4ro.legalconsultation.repository.DocumentConsolidatedRepository;
 import com.code4ro.legalconsultation.repository.DocumentMetadataRepository;
 import com.code4ro.legalconsultation.repository.DocumentNodeRepository;
+import com.code4ro.legalconsultation.repository.PdfHandleRepository;
 import com.code4ro.legalconsultation.util.CommentFactory;
 import com.code4ro.legalconsultation.util.DocumentNodeFactory;
 import com.code4ro.legalconsultation.util.PdfFileFactory;
@@ -18,6 +19,7 @@ import org.junit.Test;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.beans.factory.annotation.Value;
 import org.springframework.http.MediaType;
+import org.springframework.mock.web.MockMultipartFile;
 import org.springframework.security.test.context.support.WithMockUser;
 import org.springframework.transaction.annotation.Transactional;
 import org.springframework.util.FileSystemUtils;
@@ -50,6 +52,8 @@ public class DocumentControllerIntegrationTest extends AbstractControllerIntegra
     private DocumentNodeFactory documentNodeFactory;
     @Autowired
     private CommentFactory commentFactory;
+    @Autowired
+    private PdfHandleRepository pdfHandleRepository;
 
     @Test
     @WithMockUser
@@ -449,6 +453,23 @@ public class DocumentControllerIntegrationTest extends AbstractControllerIntegra
 
         mvc.perform(delete("/api/documents/" + uuid.toString()))
                 .andExpect(status().isNotFound());
+    }
+
+    @Test
+    @WithMockUser
+    @Transactional
+    public void addPdf() throws Exception {
+        DocumentConsolidated consolidated = saveSingleConsolidated();
+
+        final String state = "abcdef";
+        final MockMultipartFile randomFile = PdfFileFactory
+                .getAsMultipart(getClass().getClassLoader());
+        
+        mvc.perform(multipart("/api/documents/" + consolidated.getId().toString() + "/pdf")
+                .file(randomFile)
+                .contentType(MediaType.APPLICATION_PDF)
+                .param("state", state))
+                .andExpect(status().isOk());
     }
 
     private DocumentConsolidated saveSingleConsolidated() {

--- a/src/test/java/com/code4ro/legalconsultation/service/DocumentServiceTest.java
+++ b/src/test/java/com/code4ro/legalconsultation/service/DocumentServiceTest.java
@@ -3,15 +3,13 @@ package com.code4ro.legalconsultation.service;
 import com.code4ro.legalconsultation.converters.DocumentConsolidatedMapper;
 import com.code4ro.legalconsultation.model.dto.DocumentConsolidatedDto;
 import com.code4ro.legalconsultation.model.dto.DocumentMetadataDto;
-import com.code4ro.legalconsultation.model.persistence.DocumentConsolidated;
-import com.code4ro.legalconsultation.model.persistence.DocumentMetadata;
-import com.code4ro.legalconsultation.model.persistence.DocumentNode;
-import com.code4ro.legalconsultation.model.persistence.User;
+import com.code4ro.legalconsultation.model.persistence.*;
 import com.code4ro.legalconsultation.service.api.CommentService;
 import com.code4ro.legalconsultation.service.impl.DocumentConsolidatedService;
 import com.code4ro.legalconsultation.service.impl.DocumentMetadataService;
 import com.code4ro.legalconsultation.service.impl.DocumentServiceImpl;
 import com.code4ro.legalconsultation.service.impl.UserService;
+import com.code4ro.legalconsultation.service.impl.pdf.PDFServiceImpl;
 import com.code4ro.legalconsultation.util.RandomObjectFiller;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -20,6 +18,7 @@ import org.mockito.Captor;
 import org.mockito.InjectMocks;
 import org.mockito.Mock;
 import org.mockito.junit.MockitoJUnitRunner;
+import org.mockito.stubbing.Answer;
 import org.springframework.data.domain.Pageable;
 
 import java.math.BigInteger;
@@ -30,6 +29,7 @@ import java.util.UUID;
 import java.util.stream.Collectors;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertEquals;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
@@ -47,6 +47,8 @@ public class DocumentServiceTest {
     private CommentService commentService;
     @Mock
     private DocumentConsolidatedMapper documentConsolidatedMapper;
+    @Mock
+    private PDFServiceImpl pdfService;
 
     @Captor
     private ArgumentCaptor<DocumentConsolidated> documentConsolidatedArgumentCaptor;
@@ -122,5 +124,29 @@ public class DocumentServiceTest {
         final DocumentConsolidated capturedDocument = documentConsolidatedArgumentCaptor.getValue();
         assertThat(capturedDocument.getAssignedUsers()).hasSize(3);
         assertThat(capturedDocument.getAssignedUsers()).containsAll(assignedUsers);
+    }
+
+    @Test
+    public void addPdf() {
+        final UUID documentId = UUID.randomUUID();
+        
+        when(documentConsolidatedService.getEntity(any(UUID.class))).thenAnswer((Answer<DocumentConsolidated>) invocationOnMock -> {
+            DocumentConsolidated documentConsolidated = new DocumentConsolidated();
+            documentConsolidated.setId(documentId);
+            return documentConsolidated;
+        });
+        final DocumentConsolidated documentConsolidated = documentConsolidatedService.getEntity(documentId);
+        
+        assertThat(documentConsolidated.getId()).isEqualTo(documentId);
+
+        when(pdfService.createPdf(any(DocumentConsolidated.class), any(), any())).thenAnswer((Answer<PdfHandle>) invocationOnMock -> {
+            PdfHandle pdfHandle = new PdfHandle();
+            pdfHandle.setOwner(documentConsolidated);
+            return pdfHandle;
+        });
+        
+        final PdfHandle pdfHandle = pdfService.createPdf(documentConsolidated, null, null);
+
+        assertThat(documentConsolidated).isEqualTo(pdfHandle.getOwner());
     }
 }

--- a/src/test/java/com/code4ro/legalconsultation/service/PdfServiceTest.java
+++ b/src/test/java/com/code4ro/legalconsultation/service/PdfServiceTest.java
@@ -1,0 +1,75 @@
+package com.code4ro.legalconsultation.service;
+
+import com.code4ro.legalconsultation.model.persistence.DocumentConsolidated;
+import com.code4ro.legalconsultation.model.persistence.PdfHandle;
+import com.code4ro.legalconsultation.repository.PdfHandleRepository;
+import com.code4ro.legalconsultation.service.impl.FilesystemStorageService;
+import com.code4ro.legalconsultation.service.impl.pdf.PDFServiceImpl;
+import com.code4ro.legalconsultation.util.PdfFileFactory;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+import org.mockito.stubbing.Answer;
+import org.springframework.web.multipart.MultipartFile;
+
+import java.io.IOException;
+import java.net.URI;
+import java.time.Instant;
+import java.util.UUID;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.*;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class PdfServiceTest {
+
+    @Mock
+    private FilesystemStorageService storageApi;
+    @Mock
+    private PdfHandleRepository pdfHandleRepository;
+
+    @InjectMocks
+    private PDFServiceImpl pdfService;
+
+    @Test
+    public void createPdf() throws IOException {
+        final UUID ownerUuid = UUID.randomUUID();
+        final DocumentConsolidated owner = new DocumentConsolidated();
+        owner.setId(ownerUuid);
+        
+        final UUID pdfHandleId = UUID.randomUUID();
+        final String state = "abcdef";
+        final MultipartFile randomFile = PdfFileFactory
+                .getAsMultipart(getClass().getClassLoader());
+        final URI randomFilePath = URI.create("file:///" + randomFile.getOriginalFilename());
+        assertThat(randomFile).isNotNull();
+        Instant timestamp = Instant.now();
+        final Integer hash = randomFile.hashCode();
+
+        when(pdfHandleRepository.save(any(PdfHandle.class))).thenAnswer((Answer<PdfHandle>) invocationOnMock -> {
+            PdfHandle mockPdfHandle = invocationOnMock.getArgument(0);
+            mockPdfHandle.setId(pdfHandleId);
+            return mockPdfHandle;
+        });
+        when(storageApi.storeFile(any(MultipartFile.class))).thenReturn(randomFilePath.toString());
+
+        final PdfHandle pdfHandle = pdfService.createPdf(owner, state, randomFile);
+
+        Instant timestampAfter = Instant.now();
+
+        verify(pdfHandleRepository).existsByHash(hash);
+
+        assertThat(pdfHandleId).isEqualTo(pdfHandle.getId());
+        assertThat(ownerUuid).isEqualTo(pdfHandle.getOwner().getId());
+        assertThat(state).isEqualTo(pdfHandle.getState());
+        assertThat(timestamp).isBefore(pdfHandle.getTimestamp());
+        assertThat(timestampAfter).isAfter(pdfHandle.getTimestamp());
+        assertThat(hash).isEqualTo(pdfHandle.getHash());
+        assertThat(randomFilePath).isEqualTo(pdfHandle.getUri());
+    }
+}

